pkg/app/ops/deploymentchaincontroller/controller.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploymentchaincontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+const (
+	// interval represents time to sync state for all deployment chain.
+	interval = time.Minute
+	// updaterWorkerNum represents the maximum number of updaters which can be
+	// run at the same time.
+	maxUpdaterWorkerNum = 10
+)
+
+type DeploymentChainController struct {
+	applicationStore     datastore.ApplicationStore
+	deploymentStore      datastore.DeploymentStore
+	deploymentChainStore datastore.DeploymentChainStore
+	// Map from deployment chain ID to the updater
+	// who in charge for the deployment chain updating.
+	updaters map[string]*updater
+
+	logger *zap.Logger
+}
+
+func NewDeploymentChainController(
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *DeploymentChainController {
+	return &DeploymentChainController{
+		applicationStore:     datastore.NewApplicationStore(ds),
+		deploymentStore:      datastore.NewDeploymentStore(ds),
+		deploymentChainStore: datastore.NewDeploymentChainStore(ds),
+		updaters:             make(map[string]*updater),
+		logger:               logger.Named("deployment-chain-controller"),
+	}
+}
+
+func (d *DeploymentChainController) Run(ctx context.Context) error {
+	t := time.NewTicker(interval)
+	defer t.Stop()
+
+	d.logger.Info("start running deployment chain controller")
+	defer d.logger.Info("deployment chain controller has been stopped")
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-t.C:
+			d.syncUpdaters(ctx)
+		}
+	}
+}
+
+func (d *DeploymentChainController) syncUpdaters(ctx context.Context) error {
+	// Remove done updater of completed deployment chain.
+	for id, u := range d.updaters {
```

Comment:
Do we have any reason for holding this `updaters` list in `DeploymentChainController` struct?
I am seeing that the `listNotCompletedDeploymentChain` is always called to return a full list so a local variable is enough, isn't that?

Reply:
Nice question 👍 Previously, I tried to avoid the pattern of `read the deployment chain object first, then manipulate the read object and write back`, that's why I tried only to store the deployment chain id and the state of the deployment chain after it been updated (using datastore update interface) to determine should I run logic to update deployment chain object in the database or not. I thought we may miss somewhere or ops could be more than one instance or such so that avoid using the `read then write` pattern may be better for us. But you're right too, simplify the controller by making it just `fetch all not finished deployment chains then update those fetched objects then write back` is simpler, and if we can ensure the ops will always be one (ensure by the design), that will be better - I'm okay to go that way too. How do you think?

Reply:
> fetch all not finished deployment chains then update those fetched objects then write back

Yes, I think this will be simpler.
From another perspective, I can see an advantage of the current way is that we can have a faster control loop for handling the currently running chains.
More specifically, we have 2 control loops:
- one with a long interval (e.g. 1m) to find newly created chains and add them to the updater list
- one with a short internal (e.g. 15s) to execute `updater.Run` on worker

What do you think?

Reply:
I can say that the two methods (read then write back and read one, update on check) only have very few differences, that is the call to those 2 methods of deployment chain object
```golang
	applicationRefs:      dc.ListAllInChainApplications(),
	deploymentRefs:       dc.ListAllInChainApplicationDeploymentsMap(),
```
to extract applicationRefs and deploymentRefs from the deployment chain object. Can say that this call does not that costly since the number of deployments in a chain can be that huge. But I like the idea of using `short internal` only for updating the known deployment chain updaters and using a different loop to control fetching unknown deployment chains. Tbh, the thing I worry about most currently is that a 1m interval check is too long for the updating based on pull model, and the deployment chain trigger/plannable logic based on the state of the chains, long wait for updating state is not good. I think I will adopt your suggestion 👍

Reply:
Ok. Let's do it.

Reply:
addressed by https://github.com/pipe-cd/pipe/pull/2986/commits/6e2d62ab5d9c71dc490ca85de805e5738e27466c, it looks better now, PTAL when you have time 🙏